### ISSUE TITLE
mut_from_ref triggers on trait definitions

### ISSUE THREAD

**Issue by ETCaton** [rust-lang/rust-clippy#10642]
### Summary

`mut_from_ref` will trigger on trait definitions, which can't possibly be unsound. Since the lint is enabled, it should catch actual implementations of the trait in the codebase when/if they exist instead.

### Lint Name

mut_from_ref

### Reproducer

I tried this code:

```rust
#![forbid(clippy::mut_from_ref)]

type UnrelatedToSelf = ();

trait MyTrait {
    fn foo(&self, bar: &mut UnrelatedToSelf) -> &mut UnrelatedToSelf;
}
```

I saw this happen:

```
error: mutable borrow from immutable input(s)
 --> src/lib.rs:7:49
  |
7 |     fn foo(&self, bar: &mut UnrelatedToSelf) -> &mut UnrelatedToSelf;
  |                                                 ^^^^^^^^^^^^^^^^^^^^
  |
note: immutable borrow here
 --> src/lib.rs:7:12
  |
7 |     fn foo(&self, bar: &mut UnrelatedToSelf) -> &mut UnrelatedToSelf;
  |            ^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#mut_from_ref
note: the lint level is defined here
 --> src/lib.rs:2:11
  |
2 | #![forbid(clippy::mut_from_ref)]
  |           ^^^^^^^^^^^^^^^^^^^^
```

I expected not to get an error.

### Version

```text
rustc 1.69.0-beta.6 (6a97551bc 2023-04-02)
binary: rustc
commit-hash: 6a97551bcbc982df3a2da1683611dc11dc1d40cc
commit-date: 2023-04-02
host: aarch64-apple-darwin
release: 1.69.0-beta.6
LLVM version: 15.0.7
```

**Comment by J-ZhengLi** [comment]
Hi~
I'm just here pointing out that maybe this lint should indeed triggers on trait definitions here, because even if can't possibly be unsound, it could still be mistakenly defined, and clippy was designed to catch programmer mistakes.

Also, about that reproducer code, as @Niki4tap mentioned in [the zulip topic](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/mut_from_ref.20triggers.20on.20trait.20definitions/near/352245743), it should be a lifetime inference confusion, 

here's his fixed example that did not trigger this lint:

```rust
#![forbid(clippy::mut_from_ref)]

type UnrelatedToSelf = ();

trait MyTrait {
    fn foo<'a, 'b>(&'a self, bar: &'b mut UnrelatedToSelf) -> &'b mut UnrelatedToSelf;
}
```

But I do think that the error output in this case was not clear enought, which could use some improvement.
